cluster/bootstrap/bootstrap.go

```diff
@@ -103,88 +106,59 @@ func (b *Bootstrapper) Do(ctx context.Context, serverPortMap map[string]int, lg
-func (b *Bootstrapper) join(ctx context.Context, servers []string, voter bool) (leader string, err error) {
-	if entSentry.Enabled() {
-		span := sentry.StartSpan(ctx, "raft.bootstrap.join",
-			sentry.WithOpName("join"),
-			sentry.WithDescription("Attempt to join an existing cluster"),
-		)
-		ctx = span.Context()
-		span.SetData("servers", servers)
-		defer span.Finish()
-	}
-
-	var resp *cmd.JoinPeerResponse
-	req := &cmd.JoinPeerRequest{Id: b.localNodeID, Address: b.localRaftAddr, Voter: voter}
-	// For each server, try to join.
-	// If we have no error then we have a leader
-	// If we have an error check for err == NOT_FOUND and leader != "" -> we contacted a non-leader node part of the
-	// cluster, let's join the leader.
-	// If no server allows us to join a cluster, return an error
-	for _, addr := range servers {
-		resp, err = b.joiner.Join(ctx, addr, req)
-		if err == nil {
-			return addr, nil
-		}
-		st := status.Convert(err)
-		if leader = resp.GetLeader(); st.Code() == codes.NotFound && leader != "" {
-			_, err = b.joiner.Join(ctx, leader, req)
-			if err == nil {
-				return leader, nil
-			}
-		}
-	}
-	return "", fmt.Errorf("could not join a cluster from %v", servers)
-}
-
-// notify notifies another server of my presence
-func (b *Bootstrapper) notify(ctx context.Context, servers []string) (err error) {
+// notify attempts to notify all nodes in remoteNodes that this server is ready to bootstrap
+func (b *Bootstrapper) notify(ctx context.Context, remoteNodes []string) (err error) {
 	if entSentry.Enabled() {
 		span := sentry.StartSpan(ctx, "raft.bootstrap.notify",
 			sentry.WithOpName("notify"),
 			sentry.WithDescription("Attempt to notify existing node(s) to join a cluster"),
 		)
 		ctx = span.Context()
-		span.SetData("servers", servers)
+		span.SetData("servers", remoteNodes)
 		defer span.Finish()
 	}
-	for _, addr := range servers {
+	for _, addr := range remoteNodes {
 		req := &cmd.NotifyPeerRequest{Id: b.localNodeID, Address: b.localRaftAddr}
-		_, err = b.joiner.Notify(ctx, addr, req)
+		_, err = b.nodeClient.Notify(ctx, addr, req)
 		if err != nil {
 			return err
 		}
 	}
 	return
 }
 
-// servers retrieves a list of server addresses based on their names and ports.
-func (b *Bootstrapper) servers(candidates []string, serverPortMap map[string]int) []string {
-	candidates = candidates[:0]
+// ResolveRemoteNodes returns a list of remoteNodes addresses resolved using addrResolver. The nodes id used are
+// taken from serverPortMap keys and ports from the values
+func ResolveRemoteNodes(addrResolver resolver.NodeToAddress, serverPortMap map[string]int) []string {
+	candidates := make([]string, 0, len(serverPortMap))
 	for name, raftPort := range serverPortMap {
-		if addr := b.addrResolver.NodeAddress(name); addr != "" {
+		if addr := addrResolver.NodeAddress(name); addr != "" {
 			candidates = append(candidates, fmt.Sprintf("%s:%d", addr, raftPort))
 		}
 	}
+	slices.Sort(candidates)
```

Comment:
So i found it was mostly necessary for tests, but it doesn't hurt that weaviate also tries to join from the node list in the same order. Rationale behind it is in test we verify that we call the nodes in the right order (to see that we're following the response to reach a leader). It can lead to a flaky test because the map order is random above, that means the array order can be different based on that.
As I added to assert expectations on the bootstrap test it started leading to flakyness. So that fixes it and I think it's a good change on weaviate as well to always have the list ordered.